src/version.rs

```diff
@@ -0,0 +1,263 @@
+    // more than one.
+    pub fn table_version_ids(&self) -> Vec<TableVersionId> {
+        self.table_versions
+            .values()
+            .filter_map(|&table_version_id| table_version_id)
+            .collect()
+    }
+}
+
+impl<'ast> VisitorMut<'ast> for TableVersionProcessor {
+    fn visit_table_table_factor(
+        &mut self,
+        name: &'ast mut ObjectName,
+        alias: Option<&'ast mut TableAlias>,
+        args: &'ast mut Option<Vec<FunctionArg>>,
+        with_hints: &'ast mut [Expr],
+    ) {
+        if let Some(func_args) = args {
+            // TODO: Support named func args for even more flexible syntax?
+            // TODO: if func_args length is not exactly 1 error out
+            if let FunctionArg::Unnamed(FunctionArgExpr::Expr(Expr::Value(
+                Value::SingleQuotedString(value),
+            ))) = &func_args[0]
+            {
+                let unresolved_name = name.to_string();
+                let resolved_ref = TableReference::from(unresolved_name.as_str())
+                    .resolve(&*self.default_catalog, &*self.default_schema);
+                let full_object_name = ObjectName(vec![
+                    Ident::new(resolved_ref.catalog),
+                    Ident::new(resolved_ref.schema),
+                    Ident::new(resolved_ref.table),
+                ]);
+                if !self.rewrite_ready {
+                    // We haven't yet fetched/triaged the table versions ids; for now just collect
+                    // all table raw versions specified.
+
+                    self.table_versions
+                        .insert((full_object_name, value.to_string()), None);
+                } else {
+                    // Do the actual name rewrite
+                    name.0.last_mut().unwrap().value =
+                        self.table_with_version(&full_object_name, value);
+                    // Void the function table arg struct to leave a clean printable statement
+                    *args = None;
+                }
+            }
+        }
+        visit_table_table_factor(self, name, alias, args, with_hints)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::data_types::TableVersionId;
+    use datafusion::sql::parser::Statement;
+    use sqlparser::ast::Statement as SQLStatement;
+    use std::ops::Deref;
+    use test_case::test_case;
+
+    use crate::datafusion::parser::DFParser;
+    use crate::datafusion::visit::VisitorMut;
+    use crate::version::TableVersionProcessor;
+
+    #[test_case(
+        "SELECT * FROM test_table";
+        "Basic select with bare table name")
+    ]
+    #[test_case(
+        "SELECT * FROM some_schema.test_table";
+        "Basic select with schema and table name")
+    ]
+    #[test_case(
+        "SELECT * FROM some_db.some_schema.test_table";
+        "Basic select with fully qualified table name")
+    ]
+    fn test_table_version_rewrite(query: &str) {
```

Comment:
I'd add a test case here for something like a CTE or a table name that doesn't reference a version at all, something like:

```sql
WITH some_cte AS (SELECT 1 AS k) SELECT t.*, k.* FROM some_schema.test_table('2022-01-01 ...') t JOIN some_cte c ON t.k = c.k
```

Reply:
Added tests for this now.